### DISCUSSION TITLE
Modeling a nonlinear spring passively actuated joint

### DISCUSSION THREAD

**Discussion by martijnhabers** [google-deepmind/mujoco#1350] [General]
Hello,

For a project I am simulating a lower limb exoskeleton, which consists of several joints. Each joint is controlled actively by a motor, including a virtual sensor measuring position and velocity. The image below gives an indication of the approximate design. The black cylinders are the actuated joints.

<p align="center"><img src="https://github.com/google-deepmind/mujoco/assets/69632183/442f7feb-b3f0-478f-ad72-e77a23f084b5" width="400" height="400"></p>

I want to change the design of the exoskeleton and add a passive joint in the ankle, to allow for inversion/eversion movement (rotation along the heel to toe axis). This new joint will be passively actuated, contrary to the other joints. 

The movement is constrained between -8 and 8 degrees. However, within the constraints, the torque provided by the spring is given by this non-linear profile.

![AIE-torque](https://github.com/google-deepmind/mujoco/assets/69632183/f6b1b763-eb8b-460c-ac87-54fedb9ef0f9)

I am working on describing this curve as a continuous set of functions, so that I can base the torque on any given position of the ankle.

I am not very familiar with the possibilities of MuJoCo but have looked through the documentation and found that you have some form of passive joints included, but I'm having trouble to see how I should implement a position dependent torque as described above.

Does anyone have an idea on how this could be implemented? Help would be greatly appreciated!

**Comment by yuvaltassa** [comment]
The easiest way to do it is to apply your force in a [passive force callback](https://github.com/google-deepmind/mujoco/blob/e0864ab7f2cc3ed7f28c3a7af2ba3670c02bab40/include/mujoco/mujoco.h#L53). Depending on whether you are using C or Python this will look slightly different (and have vastly different effect on step time (negligible for the former, a lot for the latter)).

Tell us about your setup and we'll give you pointers.

**Reply by martijnhabers** [reply]
We are using Python. However, MuJoCo is more of an alternative to Gazebo for simulating/visualizing the movement of the exoskeleton, so how big of an impact would including this passive force callback be? Depending on how severe the slow down is I think we will try it in Python. If it proves to be too slow, we will look for alternatives.

**Reply by yuvaltassa** [reply]
Try and you'll find out 🙂 

You can see how to do it in `python/mujoco/bindings_test.py`, there's a section there that tests callbacks.
